blakserv/ccode.c

```diff
@@ -2344,3 +2350,199 @@ blak_int C_MinigameStringToNumber(int object_id,local_var_type *local_vars,
 	
 	return ret_val.int_val;
 }
+
+/**
+ * PointInPoly: Returns True if point (fx, fy) lies inside the given polygon.
+ */
+static bool PointInPoly(int fx, int fy, const server_polygon &poly)
+{
+   if (poly.num_vertices < 3)
+      return False;
+
+   bool inside = False;
+   for (int i = 0, j = poly.num_vertices - 1; i < poly.num_vertices; j = i++)
+   {
+      int xi = poly.vertices_x[i], yi = poly.vertices_y[i];
+      int xj = poly.vertices_x[j], yj = poly.vertices_y[j];
+
+      int dy = yj - yi;
+      if (dy == 0)
+         continue;
+
+      if ((yi > fy) != (yj > fy))
+      {
+         double x_int = (double) (xj - xi) * (double) (fy - yi) / (double) (yj - yi) + xi;
+
+         if (fx < x_int)
+            inside = !inside;
+      }
+   }
+   return inside;
+}
+
+/**
+ * C_PointInSector: Tests whether a fine-grained room coordinate falls inside any polygon of sectors
+ * matching given ID(s).
+ * Expects 6 params: room data, row, col, fine-row (fr), fine-col (fc), and an ID or list of IDs.
+ */
+blak_int C_PointInSector(int object_id, local_var_type *local_vars, int num_normal_parms, parm_node normal_parm_array[],
+                         int num_name_parms, parm_node name_parm_array[])
+{
+   val_type room_val, row_val, col_val, fr_val, fc_val, id_list_val;
+   std::vector<int> sector_ids;

```

Comment:
Declaring a vector can allocate some memory, so it's better to declare it right before it's used.